# README for 210208_imputation
The premise of this analysis was that we could:
1. Select a subset of the 919 individuals with the least amount of missing data and impute them with BEAGLE vesion 5.
2. Use the imputed dataset to serve as a reference panel for the remainder of the dataset.

There were some minor issues with formatting. Notably, the genotype allele separator in the VCF files is the forward slash (/), but the BEAGLE output separates the genotype alleles with the pipe symbol (|). It also lacks other output that is present in a typical VCF file like the depth of the variant, etc which are separated by a colon (:).  The [reformat_imputed_vcf.sh](reformat_imputed_vcf.sh) script was written to turn the pipe back to the foward slash so that the [normalize.awk](normalize.awk) script would work. The [normalize.awk](normalize.awk) script was also modified so that the genotype field (GT) would be written to the output because it didn't initially work like the original because of the absence of the other fields.

## Imputing with a reference panel
The script [filter_with_vcftools_by_snp_pos.sh](filter_with_vcftools_by_snp_pos.sh) was used to filter the VCF files by position. We were going to try using the SNP ID that we added using the Add_ID_to_VCF.py script, but the count starts from 0 at the beginning of each scaffold/chromosome, so there are some duplicates. As a result, I think it's better to use the combination of scaffold name + position (in basepairs) to filter. The positions to keep are located in the [snp_pos_to_keep.txt](snp_pos_to_keep.txt) file. The script [run_normalize_awk_for_filtered_imputed.sh](run_normalize_awk_for_filtered_imputed.sh) uses the AWK script [normalize_modified_for_imputed_snps.awk](normalize_modified_for_imputed_snps.awk) to convert the imputed VCF files into a TSV file. The script [run_filter_snps_and_make_wide_format_by_snp_pos.sh](run_filter_snps_and_make_wide_format_by_snp_pos.sh) then launches the R script [filter_snps_and_make_wide_format.R](filter_snps_and_make_wide_format.R) to convert the TSV file into a CSV file which can be used for downstream analyses (e.g., GWAS).

The script [impute_with_beagle_with_ref_and_filt_by_snp_pos.sh](impute_with_beagle_with_ref_and_filt_by_snp_pos.sh) was used to impute with a reference panel. The reference panel was selected based on having the least amount of missing data and was independently imputed before being used as a reference panel because BEAGLE won't permit any missing data in the reference panel. After imputation, the [reformat_imputed_vcf_filt_by_pos.sh](reformat_imputed_vcf_filt_by_pos.sh) script was used to replace the pipe character "|" with the forward slash "/" as described above.
