# README

**STRUCTURE analysis**<br>
The [emulateStructurePlots.py](emulateStructurePlots.py) script was written to mimic the outputs of the software program [STRUCTURE](https://web.stanford.edu/group/pritchardlab/structure.html) by [Pritchard et al. (2000)](https://academic.oup.com/genetics/article/155/2/945/6048111?login=true). The main reason for not simply using the image files generated by STRUCTURE is so that we could add custom labels beneath each _K_ population instead of doing that step in Powerpoint. The input should be a CSV file using text output from STRUCTURE.

To use this script on the command line, type: ```python emulateStructurePlots.py sys.argv[1] sys.argv[2]``` where ```sys.argv[1]``` is the CSV file and ```sys.argv[2]``` is the desired output (PNG format).

**Analysis of Molecular Variance (AMOVA)**<br>
The Rscript [AMOVA.R](AMOVA.R) conducts the Analysis of Molecular Variance analysis. The R script uses the [```poppr.amova()```](https://search.r-project.org/CRAN/refmans/poppr/html/poppr.amova.html) function from the [_poppr_](https://cran.r-project.org/web/packages/poppr/index.html) R package. We used the "farthest neighbor" algorithm because it is the most strict option. The ```poppr.amova()``` function gives users the choice between the [_ade4_](https://cran.r-project.org/web/packages/ade4/index.html) implementation or the [_pegas_](https://cran.r-project.org/web/packages/pegas/index.html) implementation. We chose to use the _ade4_ implementation. To correct for non-Euclidean distance, we used the ```quasieuclid``` correction. We chose this over the ```lingoes``` or ```cailliez``` correction methods because ```quasieuclid``` does not introduce a modification of the original distances like the other two methods. The [AMOVA.R](AMOVA.R) script is launched using the shell script [run_AMOVA.sh](run_AMOVA.sh).

**Mantel test**<br>
The Mantel test was conducted using the R script ```mantel_test.R```. The script uses the implementation of the Mantel test in the R package [_ade4_](https://cran.r-project.org/web/packages/ade4/index.html). Geographical and genetic distances are loaded into the R environment from their own CSV files (geo_distances.csv for geographic distances and genetic_distances.csv for genetic distances). Both tables need to be converted into an object of ```dist``` class before the ```mantel.rtest()``` can be used. From a practical purpose, this modifies the original values but it appears to change them all equally.

**Prep files for input into Arlequin**<br>
The script [convert_csv_to_arlequin_input_format.py](convert_csv_to_arlequin_input_format.py) was written to convert the SNP matrix from the comma-separated value (CSV) format to Arlequin input format. Arlequin requires somewhat unique formatting. For a diploid individual, each locus requires two lines. There needs to be a name for the haplotype, the number of individuals with that haplotype, followed by the SNP calls. The second line only need the SNP calls. For homozygous calls (0 or 1), the ref (0) or alt (2) calls are used for both lines.

The basic code for running the script is:<br>
```python convert_csv_to_arlequin_input_format.py BassLake_snp_matrix.csv BassLake_arlequin_input.arp```<br>
The script needs to be run for each lake.

The input files are:
* BassLake_snp_matrix.csv
* ClearwaterRiver_snp_matrix.csv
* DahlerLake_snp_matrix.csv
* DeckerLake_snp_matrix.csv
* GarfieldLake_snp_matrix.csv
* MudhenLake_snp_matrix.csv
* NecktieRiver_snp_matrix.csv
* OttertailRiver_snp_matrix.csv
* PhantomLake_snp_matrix.csv
* Plantagenet_snp_matrix.csv
* ShellLake_snp_matrix.csv
* UpperRiceLake_snp_matrix.csv
* ZizaniaAquatica_snp_matrix.csv

They were created by opening the original SNP matrix (with rows=SNPs and columsn=individuals) in Excel and separating the SNP data accrording to which lake/river they come from. The data were then transposed (so that rows=individuals and columns=SNPs) as required for Arlequin format.<br>

Some post-processing is required at the moment. For example, Arlequin format requires 2 rows for each haplotype, but only the first one needs a name and number of occurrence. So, I need to find a way to skip those first two columns/fields for the second row to ensure that the loci align properly. There are also extra loci for the last individual and I'm not quite sure why. <br>

I am considering each individual to have its own haplotype. We would likely see many shared haplotypes if we were using imputed data, but for now I haven't tried to reduce the number of haplotypes. I think there is too much missing data for the combination of haplotypes to be possible.

Arlequin can be difficult to work with sometimes. Here are the settings that I used to make it work:
* Each lake in its own group
    * e.g., Group 1 = Bass Lake, Group 2 = Clearwater River
* Project information (all of this is greyed out)
    * Filename: C:/Users/haasx092/Documents/wild_rice/genetic_diversity/Arlequin_files/natural_stands_fst.arp
    * Ploidy: Genotypic data
    * Gametic Phase: Unknown
    * Dominance: Codominant
    * Data type: DNA
    * Recessive allele: null
    * Locus separator: WHITESPACE
    * Missing data: ?
* Population comparisons
    * :heavy_check_mark: Compute pairwise FST
* Genetic distance settings
    * :heavy_check_mark: Slatkin's distance
    * :heavy_check_mark: Reynold's distance
    * :heavy_check_mark: Compute pairwise differences (pi)
    * No. of permutations: 100
    * Significance level: 0.05
    * Use conventional F-statistics (haplotype frequencies only)
    * Pairwise difference
    * Gamma a value: 0 
* Population differentiation
    * :heavy_check_mark: Exact test of population differentiation
    * Exact test settings
        * No. of steps in Markov chain: 100,000
        * No. of dememorization steps: 10,000
        * :heavy_check_mark: Generate histogram and table
        * Significance level: 0.05
* Project Wizard
    * Data type: FREQUENCY
    * Genotypic data: NOT CHECKED
    * Known gametic phase: GREY AND NOT CHECKED
    * Recessive data: GREY AND NOT CHECKED
    * Controls:
        * No. of samples: 2
        * Locus separator: WHITESPACE
        * Missing data: ?
    * Optional sections
        * Include haplotype list: UNCHECKED
        * Include distance matrix: UNCHECKED
        * Include genetic structure: UNCHECKED
